### DISCUSSION TITLE
Which user role allows a user to manage their account via the API?

### DISCUSSION THREAD

**Discussion by Nickk4** [keycloak/keycloak#23319] [Q&A]
We use KeyCloak 21.0.1 and would like to allow certain user actions in our application that affect KeyCloak: e.g., allow users to request a Reset password link or to de-activate their account via our application. So a user does these actions in our application, which then calls upon the KeyCloak API to process them.

We have it working in the case the user is a realm-admin. But a "regular" user does not seem to have the rights (gets a Permission denied error from KeyCloak).

What user role do we need to give users to allow them to the accounts associated to managing their account? 
It works when we give users the role of `[real-management] manage-users` but then users can also do the actions for other user accounts, besides their own account. This is obviously not desirable. 
What is the appropriate user role for this use case?

**Comment by ssilvert** [comment]
Check out the roles on the `account` client.

If you want all users to have these rights, add the roles to `default-roles-<realmname>`

**Reply by suvrat-joshi** [reply]
I created a new user and added all the roles belonging to the `account` client to that user. However, I am still not able to call the `reset-password-email` endpoint and also to deactivate itself via API. Below is a snap of all the roles I have given to this user.

![image](https://github.com/keycloak/keycloak/assets/10699092/58962120-8ef6-496c-ab7d-d03e6abfb1d7)

But whenever I make the API request, I get a `403` response status and the response body of `{ "error": "unknown_error" }`. Here are the APIs calls I am making for sending reset-password-email and to deactivate the user. It works fine when I assign the user the role of  `[real-management] manage-users` or `[real-management] realm-admin`.

**Reset Password Email**

```
PUT <keycloak>/admin/realms/<realm>/users/<userId>/reset-password-email?client_id=<clientId>&redirect_uri=http://localhost:3000/
```

**Deactivate the User**

```
PUT <keycloak>/admin/realms/<realm>/users/<userId>

Body: application/json
{
    "enabled": true
}
```

I am also sending the access token in the authorization header after signing in using `authorization_code` grant type. Not sure if we need to use a different API when making updates on self account only. But I couldn't find thing related to it in the docs. This is the REST API docs I followed https://www.keycloak.org/docs-api/22.0.1/rest-api/index.html#_overview

**Reply by Nickk4** [reply]
@ssilvert Suvrat is helping me out, so the follow up question above still relates to the OP. Hope you can help!

**Reply by baongan9400** [reply]
I have the same issue and am waiting for the answer too.

**Reply by nilsriga** [reply]
me too

**Comment by danielFesenmeyer** [comment]
Afaik, you can not use the roles of the account client on the **Admin API**. This makes sense because you might want to restrict access to the Admin API in a different way as "self-service" functionality like the **Account API**.

You can find some information on the "Account API" in [this discussion](https://github.com/keycloak/keycloak/discussions/10796). It should allow to change user data with account client roles.
Regarding on the current Account API functionality, it's probably it's easiest to have a look at the code, located here: https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java

**Reply by ace130-github** [reply]
In this discussion, as well as in the linked thread, it is claimed that the **Account API** can be used (in production, I assume). On the other hand, I see no commitment on the maintainers' side that this API is officially supported (ie reasonably stable etc). Nor is there any official documentation of the API, as far as I can see.

So relying on this API for production is risky, to put it mildly.

On the other hand, accessing user account information (read/write) is a very common requirement in most types of applications. 

So what's the suggested way to access user account information in a productive application secured by Keycloak? Is it a redirect to/ integration of the Account console, in connection with theming?

How do others solve this challenge?

**Reply by darius-m** [reply]
One approach would be to have a service account for the OIDC client that has the `manage-users` permission and fetches the user information and sends the update requests in the user's stead. The downside of this approach is that the service has pretty broad user management permissions, which could be dangerous if it becomes compromised.

**Reply by ace130-github** [reply]
Do I understand you correctly: You are suggesting *not* to use the **Account API**, but the **Admin API** which *is* documented officially?

**Reply by darius-m** [reply]
Yes, in my approach you would use the admin API, but you would not grant permissions to the users themselves, but instead give them to a service account. You would also be able to (theoretically) restrict API access to specific IPs, since you know where the requests from the service account come from, instead of exposing the admin API publicly. I was simply suggesting a possible implementation option if someone needs it, but if it was too far out of scope for the original question, you can ignore it.

Regarding the question about accessing account information, I'm not entirely sure what you meant - is it read access or write / update access? Read access should be easily achievable using the protocol mappers, that export user information in the form of attributes in various tokens. Write access (or executing operations on the user's behalf, as per the original question) is more complicated, and you will likely need to access the admin API, or the account API, as @danielFesenmeyer mentioned (though it's yet to be properly documented).
